### DISCUSSION TITLE
when and how the trainer or module move the data to gpu?

### DISCUSSION THREAD

**Discussion by FutureWithoutEnding** [Lightning-AI/pytorch-lightning#13713] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
Automatically move data to GPU is fine.
But, not knowing what moves and what doesn't, When the batch in ```training_step(self, batch, batch_idx)``` is a dict, no movement will occur.

Also don't know when to move and how to move to GPU?
I didn't find where to implement in the package!!

**Comment by akihironitta** [comment, marked as answer]
Hi @FutureWithoutEnding

**When the batch is transferred to device**
I think the pseudocode in the docs explains when a batch is transferred to each device very well:
https://pytorch-lightning.readthedocs.io/en/1.6.5/common/lightning_module.html#hooks

**What is transferred to device**
See the list of supported data structures in the documentation. If you have custom data structure, you need to override this hook in your `LightningModule`:
https://pytorch-lightning.readthedocs.io/en/1.6.5/common/lightning_module.html#transfer-batch-to-device

Also, if you need to manually transfer tensors to device, you can utilise `self.device` so that your code stays hardware-agnostic. `your_tensor.to(self.device)`:
https://pytorch-lightning.readthedocs.io/en/1.6.5/accelerators/gpu.html#init-tensors-using-type-as-and-register-buffer

**Reply by FutureWithoutEnding** [reply]
thank you very much!
I need to read the documentation properly.
